### ISSUE TITLE
duplicate symbol errors when using MotionFirebase and GeoFire

### ISSUE THREAD

**Issue by theCrab** [colinta/motion-firebase#25]
@colinta I am getting this error and am not able to move with ought nuking the whole motion-firebase sweetness from my code. I know this has to do with versions of Firebase that this two libs support. How do you deal with keeping versions updated?

**Comment by theCrab** [comment]
Its a long stack but here is what the end of it looks like

```sh
    /Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-firebase-3.3.0/lib/vendor/Firebase.framework/Firebase(FClock.o)
duplicate symbol _OBJC_IVAR_$_FOffsetClock._clock in:
    ./vendor/Pods/Firebase/libFirebaseStatic.a(FClock.o)
    /Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-firebase-3.3.0/lib/vendor/Firebase.framework/Firebase(FClock.o)
duplicate symbol _OBJC_CLASS_$_FOffsetClock in:
    ./vendor/Pods/Firebase/libFirebaseStatic.a(FClock.o)
    /Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-firebase-3.3.0/lib/vendor/Firebase.framework/Firebase(FClock.o)
duplicate symbol _OBJC_METACLASS_$_FOffsetClock in:
    ./vendor/Pods/Firebase/libFirebaseStatic.a(FClock.o)
    /Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-firebase-3.3.0/lib/vendor/Firebase.framework/Firebase(FClock.o)
duplicate symbol _OBJC_CLASS_$_FSystemClock in:
    ./vendor/Pods/Firebase/libFirebaseStatic.a(FClock.o)
    /Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-firebase-3.3.0/lib/vendor/Firebase.framework/Firebase(FClock.o)
duplicate symbol _OBJC_METACLASS_$_FSystemClock in:
    ./vendor/Pods/Firebase/libFirebaseStatic.a(FClock.o)
    /Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-firebase-3.3.0/lib/vendor/Firebase.framework/Firebase(FClock.o)
ld: 664 duplicate symbols for architecture i386
clang: error: linker command failed with exit code 1 (use -v to see invocation)
rake aborted!
Command failed with status (1): [/Applications/Xcode.app/Contents/Developer...]
/Library/RubyMotion/lib/motion/project/builder.rb:367:in `build'
/Library/RubyMotion/lib/motion/project/app.rb:78:in `build'
/Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/gems/motion-cocoapods-1.7.8/lib/motion/project/cocoapods.rb:53:in `build_with_cocoapods'
/Library/RubyMotion/lib/motion/project/template/ios.rb:68:in `block (2 levels) in <top (required)>'
/Library/RubyMotion/lib/motion/project/template/ios.rb:186:in `block in <top (required)>'
/Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/bin/ruby_executable_hooks:15:in `eval'
/Users/thecrab/.rvm/gems/ruby-2.2.1@motion-ios/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => build:simulator
(See full trace by running task with --trace)
```

and here is what my `Podfile.lock` looks like after installing GeoFire. Am sure GeoFire is breaking it.
```sh
PODS:
  - AFNetworking (2.5.4):
    - AFNetworking/NSURLConnection (= 2.5.4)
    - AFNetworking/NSURLSession (= 2.5.4)
    - AFNetworking/Reachability (= 2.5.4)
    - AFNetworking/Security (= 2.5.4)
    - AFNetworking/Serialization (= 2.5.4)
    - AFNetworking/UIKit (= 2.5.4)
  - AFNetworking/NSURLConnection (2.5.4):
    - AFNetworking/Reachability
    - AFNetworking/Security
    - AFNetworking/Serialization
  - AFNetworking/NSURLSession (2.5.4):
    - AFNetworking/Reachability
    - AFNetworking/Security
    - AFNetworking/Serialization
  - AFNetworking/Reachability (2.5.4)
  - AFNetworking/Security (2.5.4)
  - AFNetworking/Serialization (2.5.4)
  - AFNetworking/UIKit (2.5.4):
    - AFNetworking/NSURLConnection
    - AFNetworking/NSURLSession
  - Firebase (2.5.1)
  - GeoFire (1.1.3):
    - Firebase (~> 2.2)
    - FirebaseOSX (~> 2.4)

DEPENDENCIES:
  - AFNetworking (~> 2.5.0)
  - GeoFire (>= 1.1)

SPEC CHECKSUMS:
  AFNetworking: 05edc0ac4c4c8cf57bcf4b84be5b0744b6d8e71e
  Firebase: 3565ea88b830f28f2c77ecc524622bc71a20d085
  GeoFire: 88535ea6f82cb0555fcea87b36b7c263a2900ae8

COCOAPODS: 0.39.0
```

Whats the quickest solution @colinta

**Comment by colinta** [comment]
Yikes, that's crazy.  Short version: I have no idea!  But you could try this: copy all the motion-firebase ruby files into your project, and then install Firebase & GeoFire via cocoapods.

motion-firebase vendors its own "snapshotted" version of Firebase, so you don't need to / can't vendor Firebase via cocoapods if you use motion-firebase.

**Comment by colinta** [comment]
Also, if you can reproduce this in a small project on github, I'll clone it and poke around.

**Comment by mcdonamp** [comment]
As commented in the GeoFire bug, I don't think that it's an issue with the versions being different, I think that because `motion-firebase` ships with a version of Firebase already included, pulling in a second version via CocoaPods will cause the duplicate symbol errors because there are literally two copies of Firebase (at any version) sitting around. I assume that the Pods .xcodeproj and whatever RubyMotion does are incompatible in this way, since at link time, they can't resolve the fact that there are two versions.

I see two solutions:
  1. Install GeoFire.framework manually, assuming that you can pick up the import from the environment (I have no idea if `motion-firebase`/RubyMotion will surface it's Objective-C dependencies--if you can do `#import <Firebase/Firebase.h>` you'll be good to go).
  1. Remove the Firebase.framework dependency from `motion-firebase` and install the dependencies yourself. Alternatively, make `motion-firebase` a Cocoapod, and have it somehow depend on the Firebase pod. I have no clue if this is possible, but both are Ruby, so there's a slim chance that things work out.

**Comment by theCrab** [comment]
@colinta how can i remove the Firebase dependency in MF? In the `Rakefile`?

**Comment by colinta** [comment]
@theCrab This is the line that causes the trouble:

https://github.com/colinta/motion-firebase/blob/master/lib/motion-firebase.rb#L20
```app.vendor_project...```

**Comment by colinta** [comment]
Ah, ok - just comment that line out, and it should "just work".

`app.vendor_project` essentially does the same thing that cocoapods does - adds .c or .m files to a RubyMotion project.

Since you are including the framework via CocoaPods, this step is unnecessary.

Alternatively, you can copy all the ruby files and then nuke the gem, like you say (that was my first recommendation!).

**Comment by theCrab** [comment]
@colinta Awesome. I was having trouble with that. Ill keep you in the loop on this.

**Comment by theCrab** [comment]
So. I nuked the `lib/vendor` dir and removed that offending line. All works fine on my fork. The tests pass. I just need to make GeoFire ship with Firebase 2.5.x or hopefully @mcdonamp can separate GeoFire and Firebase install processes so its manually added. I like magic but not when I have to perform the trick, I think its a good idea. What do you think? :)

**Comment by theCrab** [comment]
@colinta if you agree I can do a PR.

**Comment by mcdonamp** [comment]
GeoFire doesn't ship with a particular version of Firebase, the podspec specifies a dependency on `~> 2.2` meaning that it will get the most up to date minor release, up to but not including `3.0`. It's likely that if you're not getting a specific version (the latest), then check your `Podfile.lock` to make sure it hasn't frozen at an old version or update the CocoaPods specs repo to ensure you have the latest versions available.

I recommend using `Geofire.framework` with the bundled Ruby version (unless there's a specific need for a certain Firebase version), as this is likely the far easier solution (it exists with no PRs to either repo and is a one click download and a one drag install).

As for changing the repo here--I think the problem is that in order to create a reasonable install path for the default case (developer just wants to use RM + Firebase, no GeoFire), there would have to be a Ruby gem wrapping the framework, so it could be gem installed appropriately. If we want to proceed with a PR, I suggest we solve the issue of easily pulling in `Firebase.framework` at the different versions from our CDN via `gem install` (or CocoaPods--it's Ruby and probably has a gem you could use), so that the default case of this repo can be solved.

**Comment by colinta** [comment]
@mcdonamp You lost me there.  I feel like there's a useful recommendation in there, but it's surrounded by word soup, and I just don't know what you mean.

@theCrab Since this is the first time this has come up, I don't want to rush and remove the framework and change the status quo; that said, I *would* like a way to easily include Geofire.framework.

I wonder, actually, if we should take the *opposite* approach - right now, I have been downloading the framework and including it in this gem.  So there's only one gem to download, and no need to even use Cocoapods.  We could have a second gem that wraps up Geofire, and make that gem dependent on this one, and have it work in a similar way.

What do you all think of that?

**Comment by mcdonamp** [comment]
Sorry about that :(

TL;DR: There is a way to include `Geofire.framework`, you just download it and include it in your XCode project (I assume there is still such a thing in Ruby Motion, though I don't know).

Making a Ruby Gem that does the same thing (wraps `Geofire.framework`, depends on `MotionFirebase`) is probably a reasonable way to go as well, and is likely more compatible with Ruby Motion, though unsupported by our team.

**Comment by colinta** [comment]
Gotcha.  In RubyMotion there is often NO use of Xcode - the call to `vendor_project` that is referred to above is the RubyMotion analog, though.  So “download it and include it” SHOULD be the exact same thing…

…but in practice, there is something strange about the GeoFire framework (not strange, just something I don’t know how to work with).  When I add `Firebase.framework` to a new project, it’s all good.  When I add GeoFire.framework, though, I get this error:

    ld: framework not found Firebase for architecture i386

I’ll see if I can find out more, but def strange.

The command causing this error is not `ld,` it’s actually clang++

    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -o "./build/iPhoneSimulator-9.3-Development/motion-geofirebase_spec.app/motion-geofirebase"  -filelist "/var/folders/r2/053q70n14lx83ry64h548n2w0000gn/T/linker-objs-list20160331-35342-1xsltpw" -arch i386 -arch x86_64 -isysroot "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator9.3.sdk" -F/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator9.3.sdk/System/Library/Frameworks -mios-simulator-version-min=9.3 -Wl,-no_pie -L/Library/RubyMotion/data/ios/9.3/iPhoneSimulator -lrubymotion-static -lobjc -licucore   -framework UIKit -framework Foundation -framework CoreGraphics -framework CoreFoundation -framework CFNetwork -framework CoreAudio -framework CoreLocation   -ObjC "/Users/colinta/Code/rubymotion/geofirebase/vendor/Firebase.framework/Firebase" -ObjC "/Users/colinta/Code/rubymotion/geofirebase/vendor/GeoFire.framework/GeoFire”


Here’s a repo that reproduces this error (frameworks bundled in there for ease of use)

https://github.com/colinta/geofirebase

**Comment by colinta** [comment]
Newest version removes the "freeze dried" frameworks.  Cocoapods FTW!
